src/main/java/org/apache/datasketches/kll/KllDirectSketch.java

```diff
@@ -0,0 +1,294 @@
+import static org.apache.datasketches.kll.KllPreambleUtil.extractLevelZeroSortedFlag;
+import static org.apache.datasketches.kll.KllPreambleUtil.extractM;
+import static org.apache.datasketches.kll.KllPreambleUtil.extractN;
+import static org.apache.datasketches.kll.KllPreambleUtil.extractNumLevels;
+import static org.apache.datasketches.kll.KllPreambleUtil.insertDyMinK;
+import static org.apache.datasketches.kll.KllPreambleUtil.insertLevelZeroSortedFlag;
+import static org.apache.datasketches.kll.KllPreambleUtil.insertN;
+import static org.apache.datasketches.kll.KllPreambleUtil.insertNumLevels;
+import static org.apache.datasketches.kll.KllSketch.SketchType.DOUBLES_SKETCH;
+import static org.apache.datasketches.kll.KllSketch.SketchType.FLOATS_SKETCH;
+
+import org.apache.datasketches.memory.MemoryRequestServer;
+import org.apache.datasketches.memory.WritableMemory;
+
+/**
+ * This class implements all the methods for the Direct (off-heap) sketches that are independent
+ * of the sketch type (float or double).
+ */
+abstract class KllDirectSketch extends KllSketch {
+  //All these members are constant for the life of this object. If the WritableMemory changes,
+  // it may require rebuilding this class
+  final boolean updatable = true;
+  WritableMemory levelsArrUpdatable;
+  WritableMemory minMaxArrUpdatable;
+  WritableMemory itemsArrUpdatable;
+
+  /**
+   * For the direct sketches it is important that the methods implemented here are designed to work dynamically
+   * as the sketch grows off-heap.
+   * @param sketchType either DOUBLE_SKETCH or FLOAT_SKETCH
+   * @param wmem the current WritableMemory
+   * @param memReqSvr the given MemoryRequestServer to request a larger WritableMemory
+   */
+  KllDirectSketch(final SketchType sketchType, final WritableMemory wmem, final MemoryRequestServer memReqSvr,
+      final MemoryValidate memVal) {
+    super(sketchType, wmem, memReqSvr);
+    levelsArrUpdatable = memVal.levelsArrUpdatable;
+    minMaxArrUpdatable = memVal.minMaxArrUpdatable;
+    itemsArrUpdatable = memVal.itemsArrUpdatable;
+  }
+
+  @Override
+  public int getK() {
+    return extractK(wmem);
+  }
+
+  @Override
+  public int getM() {
+    return extractM(wmem);
+  }
+
+  @Override
+  public long getN() {
+    return extractN(wmem);
+  }
+
+  @Override
+  public byte[] toUpdatableByteArray() {
+    final int bytes = (int) wmem.getCapacity();
+    final byte[] byteArr = new byte[bytes];
+    wmem.getByteArray(0, byteArr, 0, bytes);
+    return byteArr;
+  }
+
+  @Override
+  double[] getDoubleItemsArray() {
+    if (sketchType == FLOATS_SKETCH) { return null; }
+    final int items = getItemsArrLengthItems();
+    final double[] itemsArr = new double[items];
+    itemsArrUpdatable.getDoubleArray(0, itemsArr, 0, items);
+    return itemsArr;
+  }
+
+  @Override
+  double getDoubleItemsArrayAt(final int index) {
+    if (sketchType == FLOATS_SKETCH) { return Double.NaN; }
```

Comment:
logic error? and a few more like this below

Reply:
This is now fixed. I had to move these methods to the leaf level, which created some duplication, but allowed throwing with a clear message.

Reply:
DONE.